packages/flutter/lib/src/material/bottom_navigation_bar.dart

```diff
@@ -323,89 +457,84 @@ class _TileIcon extends StatelessWidget {
-  }) : super(key: key);
+    @required this.selectedFontSize,
+    @required this.unselectedFontSize,
+    @required this.showSelectedLabels,
+    @required this.showUnselectedLabels,
+  }) : assert(colorTween != null),
+       assert(animation != null),
+       assert(item != null),
+       assert(selectedFontSize != null),
+       assert(unselectedFontSize != null),
+       assert(showSelectedLabels != null),
+       assert(showUnselectedLabels != null),
+       super(key: key);
 
   final ColorTween colorTween;
   final Animation<double> animation;
   final BottomNavigationBarItem item;
+  final double selectedFontSize;
+  final double unselectedFontSize;
+  final bool showSelectedLabels;
+  final bool showUnselectedLabels;
 
   @override
   Widget build(BuildContext context) {
-    return Align(
-      alignment: Alignment.bottomCenter,
-      heightFactor: 1.0,
-      child: Container(
-        margin: const EdgeInsets.only(bottom: _kBottomMargin),
-        child: DefaultTextStyle.merge(
-          style: TextStyle(
-            fontSize: _kActiveFontSize,
-            color: colorTween.evaluate(animation),
-          ),
-          // The font size should grow here when active, but because of the way
-          // font rendering works, it doesn't grow smoothly if we just animate
-          // the font size, so we use a transform instead.
-          child: Transform(
-            transform: Matrix4.diagonal3(
-              Vector3.all(
-                Tween<double>(
-                  begin: _kInactiveFontSize / _kActiveFontSize,
-                  end: 1.0,
-                ).evaluate(animation),
-              ),
-            ),
-            alignment: Alignment.bottomCenter,
-            child: item.title,
+    Widget text = DefaultTextStyle.merge(
+      style: TextStyle(
+        fontSize: selectedFontSize,
+        color: colorTween.evaluate(animation),
+      ),
+      // The font size should grow here when active, but because of the way
+      // font rendering works, it doesn't grow smoothly if we just animate
+      // the font size, so we use a transform instead.
+      child: Transform(
+        transform: Matrix4.diagonal3(
+          Vector3.all(
+            Tween<double>(
+              begin: unselectedFontSize / selectedFontSize,
+              end: 1.0,
+            ).evaluate(animation),
           ),
         ),
+        alignment: Alignment.bottomCenter,
+        child: item.title,
       ),
     );
-  }
-}
-
-class _ShiftingLabel extends StatelessWidget {
-  const _ShiftingLabel({
-    Key key,
-    @required this.animation,
-    @required this.item,
-  }) : super(key: key);
 
-  final Animation<double> animation;
-  final BottomNavigationBarItem item;
+    if (!showUnselectedLabels && !showSelectedLabels) {
+      // Never show any labels.
+      text = Opacity(
+        alwaysIncludeSemantics: true,
+        opacity: 0.0,
+        child: text,
+      );
+    } else if (!showUnselectedLabels) {
```

Comment:
Could you combine this with the next case, and just set the opacity with an inline conditional? Seems like it would make it easier to tell what is different between the two cases.

Reply:
You're right that the two cases could be merged. I found the code easy to read as it was written